### DISCUSSION TITLE
`formatter.formatChanged()`: what are the callback args?

### DISCUSSION THREAD

**Discussion by dakur** [tinymce/tinymce#8571] [General]
In docs, there is a mention about `formatter.formatChanged()` function. Second argument is `callback` but I have just blurry idea about what is passed into it. Docs say:

> `callback (Function)` - Callback with state and args when the format is changed/toggled on/off.

In typescript declarations, I can read:

```ts
type FormatChangeCallback = (state: boolean, data: {
    node: Node;
    format: string;
    parents: Element[];
}) => void;
```

However this still didn't decrease my confusion. First, it seems to me from testing that two meanings of changing format are mixed together in this event listener:
- caret has moved to a node which has different format applied (e.g. I was on text in `h3` tag and clicked on text in `h4` tag) – the callback is triggered
- new format has applied on selected content (e.g. I was on text in `h3` tag and clicked `Format > Block format > Heading 4` so `<h3>` changed to `<h4>`) – the callback is triggered

Second, I've tried to `console.log()` the event when doing aforementioned actions in the editor and I'm not pretty sure what these variables actually contain:
- what is the `state`? state of what?
- what is the `data.format`? format of what?

Can someone give me a hint please? Basically, I'm trying to remove `id` attribute from the tag if it has just changed from heading to anything else. But I'd like to understand the callback as I've came across it also earlier.

**Comment by dakur** [comment]
Any ideas, please?

**Comment by TheSpyder** [comment]
`state` is a boolean, set based on whether the format is active or not.
`format` is the name of the format the callback is listening to, the same as the first argument to `formatChanged`, we added this so a single callback function can be used to handle multiple format changed listeners.
https://github.com/tinymce/tinymce/blob/8309d995d472f6dd4123211cb2a55fb93277e58a/modules/tinymce/CHANGELOG.md#L617

**Comment by FirstWhack** [comment]
How do I cleanup (de-register) a callback which was registered via `formatChanged`?
